### PR TITLE
feat(stdlib): add vim.base64 module

### PR DESCRIPTION
**Pull request by gpanders**
Add base64 encode() and decode() functions to a vim.base64 module.

---

Adding this as a prerequisite for a builtin OSC 52 clipboard provider.

TODO:

- [x] Documentation

### CODE REVIEW

File: test/functional/lua/base64_spec.lua [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+      [[
+        Call me Ishmael. Some years ago—never mind how long precisely—having little or no money in
+        my purse, and nothing particular to interest me on shore, I thought I would sail about a
+        little and see the watery part of the world.
+      ]],
+      [[
+        It is a truth universally acknowledged, that a single man in possession of a good fortune,
+        must be in want of a wife.
+      ]],
+      'Happy families are all alike; every unhappy family is unhappy in its own way.',
```
Comment (bfredl) [inline test/functional/lua/base64_spec.lua:32]: as an additional check, I would also `eq([[the base64 data]], encode(v))` at least for the shortest one.

Comment (gpanders) [reply test/functional/lua/base64_spec.lua:32]: Added one more test case with an explicit check on the encoded output.

File: src/nvim/lua/base64.c [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+
+  int acc = 0;
+  int acc_len = 0;
+  size_t out_i = 0;
+  size_t src_i = 0;
+  int leftover_i = -1;
+
+  for (; src_i < src_len; src_i++) {
+    const char c = src[src_i];
+    const uint8_t d = char_to_index[c & 0xFF];
```
Comment (ii14) [inline src/nvim/lua/base64.c:149]: `c` is already `char`
```suggestion
    const uint8_t d = char_to_index[(uint8_t)c];
```

Comment (gpanders) [reply src/nvim/lua/base64.c]: 452da56afa706a8da2f0a2948014c8fb6ebb515d

File: src/nvim/base64.c
Comment (justinmk) [inline src/nvim/base64.c:1]: is there an upstream source for this (which we may want to follow)?

File: src/nvim/lua/base64.c [hunk 3]
```diff
@@ -0,0 +1,65 @@
+#include <assert.h>
+#include <lauxlib.h>
+#include <lua.h>
+
+#include "nvim/base64.h"
+#include "nvim/lua/base64.h"
+#include "nvim/memory.h"
+
+static int nlua_base64_encode(lua_State *L)
```
Comment (justinmk) [inline src/nvim/lua/base64.c:9]: 👍  the `L` convention is very common in Lua so we should change our other code to use that too.

File: src/nvim/lua/stdlib.c [hunk 4]
```diff
@@ -606,6 +607,10 @@ void nlua_state_add_stdlib(lua_State *const lstate, bool is_thread)
     lua_pushcfunction(lstate, &nlua_iconv);
     lua_setfield(lstate, -2, "iconv");
 
+    // vim.base64
+    luaopen_base64(lstate);
+    lua_setfield(lstate, -2, "base64");
```
Comment (justinmk) [inline src/nvim/lua/stdlib.c:612]: We already have `vim.json` and `vim.pack`, so yeah, it's hard to say whether we should introduce e.g. `vim.codec` ([python module](https://docs.python.org/3/library/codecs.html) for reference).

**Pull request merged**